### ISSUE TITLE
Status of pip wheels with _GLIBCXX_USE_CXX11_ABI=1

### ISSUE THREAD

**Issue by elias-work** [pytorch/pytorch#51039]
#17492 shows the history of this issue but it has been closed and buried for a long time. Torch pip wheels are compiled with _GLIBCXX_USE_CXX11_ABI=0, resulting in incompatibility with other libraries.

Is there any sort of status on this?

(Personal motivation: Our project takes many hours to compile because of torch + dependencies + libraries that use torch. We can't link with the libraries from the pip wheel because of this, and forcing everything else to use _GLIBCXX_USE_CXX11_ABI=0 is a huge headache itself that will only result in more problems over time. Having _GLIBCXX_USE_CXX11_ABI=1 pip wheels would drastically simplify our, and I'm sure many others', builds).

**Comment by bdhirsh** [comment]
marking hi-pri due to user activity in the previous issue (which still seems unresolved)

**Comment by malfet** [comment]
I think the core problem is that _GLIBCXX_USE_CXX11_ABI=1 is not supported by compiler that is recommended by PEP513, see  https://github.com/pypa/manylinux

**Comment by seemethere** [comment]
This is actually a bug that's been reported for RHEL / CentOS 7 (devtoolset-7): https://bugzilla.redhat.com/show_bug.cgi?id=1546704

Unfortunately if we want to provide `manylinux` support, which are binaries built on CentOS 7, then we can't support `_GLIBCXX_USE_CXX11_ABI=1` in our wheels.

**Comment by elias-work** [comment]
My understanding is this is mostly an issue when you need to use torch with both python and C. Is there a recommended way to get this setup work other than by building from source? It sounds like the only options are
* Special non-standard _GLIBCXX_USE_CXX11_ABI=1 wheels
* Conda package? (I heard some people say it has this flag, but I haven't tested yet).
* Build from source
* Install and use both libtorch and Python package (don't think it will work for us)

**Comment by mattip** [comment]
To add another clarification on top of the previous clarifications: even moving to the manylinux2014 spec would not work around the hard limitation mentioned above. The upcoming [PEP 600 perennial multibuild](https://github.com/pypa/manylinux/issues/542) standard might be the first that would support this, since it would move past CentOS7 (so obviously the wheels would not be expected to work on CenOS7). All the infrastructure is in place for a manylinux_2_24 standard based on glibc 2.24 (debian stretch). The work needed to build the docker images is currently in progress, here is one of the active tracking issues pypa/manylinux#877

**Comment by EricSteinberger** [comment]
What's the best workaround at the moment? I'm not familiar with building pytorch from source, but will of course dive into it if needed. I'm exporting a C++ .so library hoping to import it in Python. Works well on mac but running into this issue on Linux machines. Any tips on workarounds?

**Comment by mattip** [comment]
Would using the conda builds solve this problem?

**Comment by EricSteinberger** [comment]
@mattip nope. Running `python -c "import torch; print(torch._C._GLIBCXX_USE_CXX11_ABI)"` with the conda version also yields `False`

**Comment by Algomorph** [comment]
Sorry if this has already been discussed in previous issues: is there any reason not to build a `manylinux`-compatible wheel set and another _GLIBCXX_USE_CXX11_ABI=1 wheel set? It seems like this could be another switch / option to choose from on the ["Get Started" Page](https://pytorch.org/get-started/locally/).

**Comment by mattip** [comment]
Wheels on PyPI must support one of the known formats, so the `_GLIBCXX_USE_CXX11_ABI=1` wheels should be hosted elsewhere, and users told specifically what flags to use with pip to force it to use an incompatible wheel.

**Comment by Algomorph** [comment]
> Wheels on PyPI must support one of the known formats, so the `_GLIBCXX_USE_CXX11_ABI=1` wheels should be hosted elsewhere, and users told specifically what flags to use with pip to force it to use an incompatible wheel.

In this case, it might make sense to serve the wheels as downloadable files from download.pytorch.org directly, similar to how the LibTorch packages are served (i.e. `C++11 Wheels` category for `Package` or something more user-friendly).

I'm not familiar with the current process of setting up wheel distribution through PyPl for the various platforms, but if there are automated ways of building the pip wheels for the three supported OS, perhaps some of these can be adapted and reused to build the C++11 wheels for each release.

I guess what remains after this is just making the changes to the website (assuming this is the chosen distribution site) and uploading the wheels (+ posting hashes to verify, perhaps ?).

Sorry if I'm being a little hand-wavy here, just trying to outline what needs to be done.

**Comment by blackliner** [comment]
Are there wheels available with D_GLIBCXX_USE_CXX11_ABI=1 ?

**Comment by Algomorph** [comment]
@blackliner, right now -- not unless you build one yourself, from source. It's not a very difficult process, just requires a long time and CPU resources. Took me about half a day all told. But then you also have to build custom wheels for torchvision and torchaudio, depending on what you're using.

**Comment by blackliner** [comment]
NVIDIA supplies the world of `aarch64` with pretty much all you need (and with the new `GLIBCXX_USE_CXX11_ABI`): https://forums.developer.nvidia.com/t/pytorch-for-jetson-version-1-7-0-now-available/72048

Very weird that `x86_64` is lacking such a variety 😞

**Comment by Algomorph** [comment]
Just noting here that the pypa/manylinux#877 is now resolved & closed. I'm not sure what the status is on the newer standards, but manylinux2014 seems to have reached end of life at this point.

**Comment by mattip** [comment]
> manylinux2014 seems to have reached end of life at this point

Centos7 that underlies manylinux2014 will [reach eol in 2024-06](https://github.com/mayeut/pep600_compliance#acceptable-distros-to-build-wheels) so it still has a bit of life left.

**Comment by andretag** [comment]
(apologies for the bump) - is there any update on this, or any workaround which does not involve building torch from source?

(personal background: I am interested in including torch model in a C++ ros package for deployment on a Jetson TX2, and I encounter linking problems similar to what is described [here](https://discuss.pytorch.org/t/libtorch-with-new-c-11-abi/39490), which lead me to this issue.)

**Comment by guoyejun** [comment]
@andretag  a dockerfile is added to support cxx11 abi build for torch wheel file, see https://github.com/pytorch/builder/blob/main/manywheel/Dockerfile_cxx11-abi, and now, we still need to build from source.

btw, the built wheel file works with manylinux_2_28.

**Comment by rgommers** [comment]
> Centos7 that underlies manylinux2014 will [reach eol in 2024-06](https://github.com/mayeut/pep600_compliance#acceptable-distros-to-build-wheels) so it still has a bit of life left.

There's no reason to not start published `manylinux_2_28` wheels though. Those will support `_GLIBCXX_USE_CXX11_ABI=1`. `manylinux_2_24` is fine to ignore (going EOL now). See https://github.com/pypa/manylinux/issues/1332, that discussed is extremely informative about which `manylinux` flavors have which GCC, support which Python versions and distros, and also touches on `_GLIBCXX_USE_CXX11_ABI`.

PyTorch 1.12.0 provides `manylinux1` and `manylinux2014` wheels, see https://pypi.org/project/torch/1.12.0/. I believe the next release should:
- drop `manylinux1`
- keep `manylinux2014` (unchanged)
- add `manylinux_2_28` wheels built with `_GLIBCXX_USE_CXX11_ABI=1`

I don't think there's an issue with providing two sets of wheels with different ABI. On distros that support `2_28`, those wheels will be preferred by Pip. And old distros will get `manylinux2014`. Packages that depend on PyTorch can/should build matching sets of wheels.

**Comment by adizhol** [comment]
This issue should go into the pyTorch C++ Extention documentation as a warning...
https://pytorch.org/tutorials/advanced/cpp_extension.html

**Comment by zhuhong61** [comment]
We have added _GLIBCXX_USE_CXX11_ABI=1 support for linux cpu wheel, based on the Ubuntu18.04 docker image https://github.com/pytorch/builder/pull/990 and Centos 8 https://github.com/pytorch/builder/pull/1023.

**Comment by CLARKBENHAM** [comment]
This is true for pytorch=1.12.1 but pytorch=2.1.0.dev20230608 seems like it's not being built with CXX11_ABI.

The package I installed today
```
$ conda list -f pytorch
# packages in environment at /root/local/miniconda:
#
# Name                    Version                   Build  Channel
pytorch                   2.1.0.dev20230608     py3.8_cpu_0    pytorch-nightly
$ python -c "import torch; print(torch._C._GLIBCXX_USE_CXX11_ABI)"
False
```

From a package installed previously:
```
$ conda list -f pytorch
# packages in environment at /root/local/miniconda:
#
# Name                    Version                   Build  Channel
pytorch                   1.12.1          cpu_py37h9dbd814_1  
$ python -c "import torch; print(torch._C._GLIBCXX_USE_CXX11_ABI)"
True
```

**Comment by nsmithtt** [comment]
Is there any update on this issue?  Is it possible to offer cxx11 ABI wheel?

**Comment by tuero** [comment]
Running into this issue as well working on a mixed python/c++ codebase which uses libtorch. It would be nice to install pytorch through pip/conda and link against that on the C++ end, but this is currently not feasible with cxx11 ABI off.

**Comment by CalaveraLoco** [comment]
Any news? Running into the same issues here. It really makes it difficult to develop hybrid solutions. Unfortunately it's nigh impossible to convert people around me to use non-stock pip packages as base.

**Comment by njzjz** [comment]
There seem to be undocumented cxx11 wheels in download.pytorch.org:

```sh
pip install "torch==2.3.0+cpu.cxx11.abi" -i https://download.pytorch.org/whl/
python -c "import torch; print(torch._C._GLIBCXX_USE_CXX11_ABI)"
True
```

However, there are two limitations: (1) no GPU wheels; (2) no correct manylinux tags (it seems to be `manylinux_2_28`).

**Comment by smanolloff** [comment]
I ran into this issue and while trying to port a project that has a C++ binary which needs to link against libtorch.
On Mac, C++ binaries are compiled **without** CXX11 ABI and linking against libtorch shipped with `pip install pytorch` work fine.
On Linux, C++ binaries are compiled **with** CXX11 ABI and the linker complains loudly. 

I one possible approach here would be to package a non-CXX11 libtorch for MacOS platforms, and a CXX11 libtorch for linux platforms (no idea about Windows)

**Comment by trns1997** [comment]
Any updates on this issue, i love the idea of torch installed via pip (`cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=python3 -c 'import torch;print(torch.utils.cmake_prefix_path)'` and linking it to my cpp project. As i mostly use conan packages it is easy to create a work around as i can simply pass cxx flag `CXXFLAGS=-D_GLIBCXX_USE_CXX11_ABI=0` to my conan profile. But since CentOS7 is EOL already, why not build with abi, that way special tweaks won't be necessary to build torch compatible projects.

**Comment by artyom-beilis** [comment]
Hello, what is the status?

I noticed when building OpenCL out of tree backend for torch 2.4 that old ABI is still in use. 

While I can easily build my own code with older ABI, I'm planning on integrating 3rd part libraries to improve performance: currently I think of two oneDNN with OpenCL support and CLBlast... Now all C++ libraries. In order to integrate them I must rebuild them with old ABI - it all complicates build and deployment.

**Comment by artyom-beilis** [comment]
One more thing: recently the C++ requirement moved to C++17 and gcc >= 9.4 on linux - isn't it a time to drop this definition?

**Comment by tengyifei** [comment]
It would be great to have C++11 ABI as the default since the newer ABI is more performant. In the XLA backend of PyTorch [1], we're offering additional C++11 ABI wheel and docker variants because the LazyTensor tracing performance is much better there.

[1]: https://github.com/pytorch/xla?tab=readme-ov-file#use-nightly-build-with-c11-abi-after-10282024

**Comment by cooopermaira** [comment]
I'm going to go ahead and assume there hasn't been any progress on this. Bummer.

**Comment by tuero** [comment]
I just tried this using a fresh pip install of torch 2.7 and can confirm I can link to it. Exciting!

**Comment by trns1997** [comment]
> Any updates on this issue, i love the idea of torch installed via pip (`cmake .. -DCMAKE_BUILD_TYPE=Release -DCMAKE_PREFIX_PATH=python3 -c 'import torch;print(torch.utils.cmake_prefix_path)'` and linking it to my cpp project. As i mostly use conan packages it is easy to create a work around as i can simply pass cxx flag `CXXFLAGS=-D_GLIBCXX_USE_CXX11_ABI=0` to my conan profile. But since CentOS7 is EOL already, why not build with abi, that way special tweaks won't be necessary to build torch compatible projects.

I'll retry this with the new version. If it works it's awesome, I will no longer need to download torch cxx tar as an external lib.
